### DISCUSSION TITLE
GSOC 2021 Ideas

### DISCUSSION THREAD

**Discussion by jneira** [haskell/haskell-language-server#1237] [Ideas]
* The [Gsoc 2021 call for ideas has started](https://discourse.haskell.org/t/gsoc-2021-call-for-ideas/1775) and it would be great that hls would be present, as in previous ones
* We are [plenty of features request](https://github.com/haskell/haskell-language-server/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+enhancement%22), but it should have the appropiate scope (and i am not sure what can be)
* Some quick ideas:
  *  Improve template haskell support: we have [many bug reports about](https://github.com/haskell/haskell-language-server/labels/type%3A%20template%20haskell%20related) 
  * REPL Support: https://github.com/haskell/haskell-language-server/issues/477
  * There are several requests about suggestions, maybe they can be grouped
  * [Tackle build tools interop through hie-bios](https://github.com/haskell/haskell-language-server/issues/1#issuecomment-698222658): `cabal show-build-info`, the stack correlated functionality
  * The performance side: use of memory for example (several issues about too)
  * ....

**Comment by Ailrun** [comment]
How about stack/cabal integration of script (non-project) cradles?

**Comment by jaspervdj** [comment]
I think we'd be okay including several of these smaller issues.  Is someone willing to summarize them so we can include them on the summer of Haskell page?  I can also help if it's a pretty literal transcription of these issues.  I'd like to get some ideas on our page over the next two weeks since it boosts Haskell.org's application to GSoC.

**Comment by jneira** [comment]
I've created a label [Eligible for gsoc 2021](https://github.com/haskell/haskell-language-server/labels/Eligible%20for%20gsoc%202021)
If we have isolated issues that can be used as base for gsoc we could label them, but if we want group several issues i would create a meta new one linking them and providing some context (and label it of course)

**Comment by jneira** [comment]
Another path to make progress could be in the renaming feature, that is almost missing in hls afaik
* There exists the retrie plugin and a issue about https://github.com/haskell/haskell-language-server/issues/282
* In hie we were using [HaRe](https://github.com/alanz/HaRe) to refactor duplicate definitions f.e.

Could both be used?, would they be alternative or complementary? is a good idea make progress in the renaming feature thanks to the next Gsoc?

//cc @pepeiborra @alanz

**Reply by pepeiborra** [reply]
The renaming feature is very doable and it would be a great contribution. The main problem is that retrie does not actually work that well in big projects, because it doesn't use the renamed AST. So the project would involve 2 steps:
1. Enhancing retrie to use the renamed AST as described in https://github.com/facebookincubator/retrie/issues/10 
2. Implementing the refactorings described in #282 including rename.

I'm happy to mentor if @xich and @alanz are busy

**Reply by jneira** [reply]
nice, i've labeled #282 as eligible for gsoc

**Comment by michaelpj** [comment]
What about some of the big-ticket LSP features that we don't have yet? e.g. semantic tokens and call hierarchies. Possibly not enough work for a full GSOC, but you could extend it to "make sure we're using all the features of the spec that we can".

**Reply by jneira** [reply]
I would love to have call hierarchy, I think thanks it would be feasible using the already included hiedb

**Reply by michaelpj** [reply]
Yeah, I think it's pretty feasible, needs some work in the `lsp` library too, but I think mostly straightforward.

**Comment by wz1000** [comment]
I posted the following call for mentors and students regarding bytecode serialiation and/or fat interface files to the ghc-devs mailing list: https://mail.haskell.org/pipermail/ghc-devs/2021-March/019633.html

**Comment by jaspervdj** [comment]
I've added a stub here so this thread gets some visibility for students: <https://github.com/haskell-org/summer-of-haskell/blob/master/content/ideas/hls.md>

**Comment by July541** [comment]
Hello, I am a student who would like to apply for GSoC. HLS is a very useful project for every developer. As a beneficiary, I hope to contribute to this project. As HLS is a complex project that beyond my ability, some technical guide may help me establish a feasible plan. I want to know where I can get some help with the details of the proposal.

**Reply by pepeiborra** [reply]
I'm not sure if this is what you had in mind, but the [IFL 2020 draft proceedings](https://www.cs.kent.ac.uk/events/2020/ifl20/ifl2020draftproceedings.pdf) contain a research paper describing some of the internals of HLS. Past GSoC students like @fendor and @bubba might be able to offer more advice!

**Reply by July541** [reply]
Thank you! I'll keep moving by your instruction.

**Reply by fendor** [reply]
My advice is to get to know the ecosystem around HLS (read the blog posts at https://mpickering.github.io/ide/index.html for an overview), start contributing to minor issues and find a pain point you want to fix. Maybe even implement a feature before submitting your proposal.

This discussion provides some ideas about existing pain points, like cabal/stack scripts, performance issues, etc...

If you find an issue you want to work on, dont be afraid to ask questions, either on the issue or on IRC freenode #haskell-language-server

**Reply by jneira** [reply]
> start contributing to minor issues and find a pain point you want to fix

Note that some issues are tagged with [first good issue](https://github.com/haskell/haskell-language-server/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)

**Reply by July541** [reply]
@fendor Thank you for your warm reply, I am trying your advice.

@jneira Thank you for your mention, I really focused on [Eligible for GSoC 2021](https://github.com/haskell/haskell-language-server/labels/Eligible%20for%20gsoc%202021) before and ignored this newer friendly tag.

**Comment by jaspervdj** [comment]
@jneira Could you email me a list of email addresses of people who would be interested in mentoring?  That way I can invite them to the GSoC dashboard.

**Reply by jneira** [reply]
Hi, i think @pepeiborra [has mentioned](https://github.com/haskell/haskell-language-server/discussions/1237#discussioncomment-456825) he could be one of them, to help to implement renaming.
Not sure about others, sorry. 
I am afraid it would not have enough time and proficiency to be a main mentor. I would try to help for sure, being officially a (secondary) mentor or not.
//cc @mpickering @wz1000 @michaelpj @Ailrun @fendor @berberman @isovector @ndmitchell

**Reply by isovector** [reply]
I'm happy to mentor

**Comment by OliverMadine** [comment]
If the proposals are still under review, I’d like to mention that I have updated some of my “draft” proposal with what I think would be a better approach to the renaming types. I also made some other minor changes.

Before I mentioned using Retrie's type forward to address the renaming issues (specifically with types),but since Retrie has limitations with larger projects, I think a combination of using HieDb references and file contents would work better. The details can be found on the draft proposal if access is still available to mentors?
